File: mobile/android/app/src/main/java/com/playbridge/sender/cast/DashParser.kt
package com.playbridge.sender.cast

import android.util.Xml
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.xmlpull.v1.XmlPullParser
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL

/**
 * Parser for MPEG-DASH manifests (.mpd).
 *
 * Extracts video Representation entries from the manifest so the UI can show the user
 * what quality tiers are available. For playback, the original MPD URL is always sent to
 * the TV player — the player handles adaptive bitrate selection natively.
 */
object DashParser {

    suspend fun parseManifest(mpdUrl: String, headers: Map<String, String>? = null): List<VideoQuality> =
        withContext(Dispatchers.IO) {
            try {
                val content = fetchUrlContent(mpdUrl, headers)
                parseXml(content, mpdUrl)
            } catch (e: Exception) {
                e.printStackTrace()
                emptyList()
            }
        }

    private fun parseXml(xmlContent: String, mpdUrl: String): List<VideoQuality> {
        val qualities = mutableListOf<VideoQuality>()

        val parser: XmlPullParser = Xml.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        parser.setInput(xmlContent.reader())

        // AdaptationSet-level fallback dimensions (some MPDs put width/height here)
        var inVideoAdaptationSet = false
        var adaptationWidth = 0
        var adaptationHeight = 0

        var event = parser.eventType
        while (event != XmlPullParser.END_DOCUMENT) {
            when (event) {
                XmlPullParser.START_TAG -> when (parser.name) {
                    "AdaptationSet" -> {
                        val mimeType   = parser.getAttributeValue(null, "mimeType")   ?: ""
                        val contentType = parser.getAttributeValue(null, "contentType") ?: ""
                        val codecs     = parser.getAttributeValue(null, "codecs")     ?: ""
                        inVideoAdaptationSet = mimeType.startsWith("video/") ||
                            contentType == "video" ||
                            VIDEO_CODECS.any { codecs.startsWith(it) }
                        adaptationWidth  = parser.getAttributeValue(null, "width")?.toIntOrNull()  ?: 0
                        adaptationHeight = parser.getAttributeValue(null, "height")?.toIntOrNull() ?: 0
                    }

                    "Representation" -> if (inVideoAdaptationSet) {
                        val bandwidth = parser.getAttributeValue(null, "bandwidth")?.toLongOrNull() ?: 0L
                        val width  = parser.getAttributeValue(null, "width")?.toIntOrNull()  ?: adaptationWidth
                        val height = parser.getAttributeValue(null, "height")?.toIntOrNull() ?: adaptationHeight
                        val codecs = parser.getAttributeValue(null, "codecs")

                        // Skip audio-only representations that sneak into video AdaptationSets
                        val isAudio = codecs != null && AUDIO_CODECS.any { codecs.startsWith(it) }

                        if (bandwidth > 0 && height > 0 && !isAudio) {
                            qualities.add(
                                VideoQuality(
                                    resolution = "${height}p",
                                    bandwidth  = bandwidth,
                                    // Always send the full MPD URL — TV player handles ABR
                                    url        = mpdUrl,
                                    codecs     = codecs
                                )
                            )
                        }
                    }
                }

                XmlPullParser.END_TAG -> when (parser.name) {
                    "AdaptationSet" -> {
                        inVideoAdaptationSet = false
                        adaptationWidth  = 0
                        adaptationHeight = 0
                    }
                }
            }
            event = parser.next()
        }

        // Deduplicate by (resolution, bandwidth) — some MPDs repeat representations
        return qualities
            .distinctBy { "${it.resolution}@${it.bandwidth}" }
            .sortedByDescending { it.bandwidth }
    }

    private fun fetchUrlContent(urlString: String, headers: Map<String, String>? = null): String {
        val resolved = resolveUrl(urlString)
        val connection = URL(resolved).openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = 10_000
        connection.readTimeout    = 10_000
        connection.instanceFollowRedirects = true

        headers?.forEach { (key, value) ->
            if (!key.equals("Range", ignoreCase = true)) connection.setRequestProperty(key, value)
        }
        if (headers?.keys?.none { it.equals("User-Agent", ignoreCase = true) } != false) {
            connection.setRequestProperty("User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
        }

        return connection.inputStream.use { BufferedReader(InputStreamReader(it)).readText() }
    }

    private fun resolveUrl(url: String): String = try {
        URI(url).toString()
    } catch (e: Exception) {
        url
    }

    private val VIDEO_CODECS = listOf("avc", "hvc", "hev", "vp8", "vp9", "av01", "dvh")
    private val AUDIO_CODECS = listOf("mp4a", "ac-3", "ec-3", "opus", "flac", "vorbis")
}

File: mobile/apple/PlayBridge Phone/PlayBridge Phone/Browser/DASHParser.swift
import Foundation

/// Parses an MPEG-DASH manifest for its video quality tiers. Port of `cast/DashParser.kt`.
/// As on Android, every tier casts the **full MPD URL** — the TV player handles ABR; the tiers
/// are informational.
enum DASHParser {
    private static let videoCodecs = ["avc", "hvc", "hev", "vp8", "vp9", "av01", "dvh"]
    private static let audioCodecs = ["mp4a", "ac-3", "ec-3", "opus", "flac", "vorbis"]

    static func variants(mpdURL: String, headers: [String: String]) async -> [VideoQuality] {
        guard let content = await StreamHTTP.fetchText(mpdURL, headers: headers),
              let data = content.data(using: .utf8) else { return [] }
        let parser = XMLParser(data: data)
        let delegate = Delegate(mpdURL: mpdURL)
        parser.delegate = delegate
        parser.parse()
        // Dedup by resolution@bandwidth, highest first.
        var seen = Set<String>()
        return delegate.qualities
            .filter { seen.insert("\($0.label)@\($0.bandwidth)").inserted }
            .sorted { $0.bandwidth > $1.bandwidth }
    }

    private final class Delegate: NSObject, XMLParserDelegate {
        let mpdURL: String
        var qualities: [VideoQuality] = []
        private var inVideoSet = false
        private var setWidth = 0, setHeight = 0

        init(mpdURL: String) { self.mpdURL = mpdURL }

        func parser(_ parser: XMLParser, didStartElement name: String, namespaceURI: String?,
                    qualifiedName: String?, attributes attr: [String: String]) {
            switch name {
            case "AdaptationSet":
                let mime = attr["mimeType"] ?? ""
                let ctype = attr["contentType"] ?? ""
                let codecs = attr["codecs"] ?? ""
                inVideoSet = mime.hasPrefix("video/") || ctype == "video"
                    || DASHParser.videoCodecs.contains(where: codecs.hasPrefix)
                setWidth = Int(attr["width"] ?? "") ?? 0
                setHeight = Int(attr["height"] ?? "") ?? 0
            case "Representation" where inVideoSet:
                let bandwidth = Int64(attr["bandwidth"] ?? "") ?? 0
                let height = Int(attr["height"] ?? "") ?? setHeight
                let codecs = attr["codecs"]
                let isAudio = codecs.map { c in DASHParser.audioCodecs.contains(where: c.hasPrefix) } ?? false
                if bandwidth > 0, height > 0, !isAudio {
                    qualities.append(VideoQuality(label: "\(height)p", bandwidth: bandwidth, url: mpdURL, codecs: codecs))
                }
            default:
                break
            }
        }

        func parser(_ parser: XMLParser, didEndElement name: String, namespaceURI: String?, qualifiedName: String?) {
            if name == "AdaptationSet" { inVideoSet = false; setWidth = 0; setHeight = 0 }
        }
    }
}
